quantum/matrix.c

```diff
@@ -92,11 +96,8 @@ static inline void setPinInputHigh_atomic(pin_t pin) {
 }
 
 static inline uint8_t readMatrixPin(pin_t pin) {
-    if (pin != NO_PIN) {
-        return readPin(pin);
-    } else {
-        return 1;
-    }
+    if ((pin == NO_PIN) || (readPin(pin) != PRESSED_KEY_PIN_STATE)) return 1;
+    if (readPin(pin) == PRESSED_KEY_PIN_STATE) return 0;
```

Comment:
```suggestion
    return 0;
```
The repeated reading of the pin should not be necessary (actually the compiler should be complaining, because the second call to `readPin(pin)` is not guaranteed to return the same result as the first, and therefore you should get the “execution reaches end of a non-void function” warning).

But probably the whole function would be easier to understand if you just wrote it like
```c
static inline uint8_t readMatrixPin(pin_t pin) {
    if (pin != NO_PIN) {
        return (readPin(pin) == PRESSED_KEY_PIN_STATE) ? 0 : 1;
    } else {
        return 1;
    }
}
```

Or maybe even
```c
static inline bool readMatrixPin(pin_t pin) {
    if (pin != NO_PIN) {
        return readPin(pin) == PRESSED_KEY_PIN_STATE;
    } else {
        return 0;
    }
}
```
and then adjust the invocations to that return value; although that might be done as a separate change.

Reply:
Agreed with this, the timing is tight, so it should match *most of the times*, but it's not guaranteed to match *all of the times*. Returning the `readPin(pin) == PRESSED_KEY_PIN_STATE` as in the second option requires more wrapping and manipulation than it's necessary, so probably going to go with the first proposal.